cerbero/build/fridge.py

```diff
@@ -0,0 +1,279 @@
+        files = self.cookbook.recipe_installed_files(recipe.name)
+        if not files:
+            m.warning('The recipe %s has no installed files. Try running all steps for the recipe from scratch' % recipe.name)
+            raise EmptyPackageError(p.name)
+
+        # Update list of installed files to make sure all of the files actually exist
+        installed_files = self.cookbook.update_installed_files(recipe.name, files)
+        paths = tar.pack_files(self.binaries_local, PackageType.DEVEL, installed_files)
+        # paths = tar.pack(self.binaries_local, devel=True, force=True, force_empty=False,
+        #                 relocatable=True)
+        p.post_package(paths, self.binaries_local)
+
+    def upload_binary(self, recipe):
+        self._ensure_ready(recipe)
+        packages = self._get_package_names(recipe)
+        fetch_packages = []
+        for p in packages.values():
+            if os.path.exists(os.path.join(self.binaries_local, p)):
+                fetch_packages.append(p)
+            else:
+                m.warning("No package was created for %s" % p)
+        self.binaries_remote.upload_binary(fetch_packages, self.binaries_local,
+                                           self.env_checksum, self.env_file)
+
+    def _get_package_names(self, recipe):
+        ret = dict()
+        # TODO: separate runtime and devel packages after setting
+        # all files needed by each recipe properly
+        #ret = {PackageType.RUNTIME: None, PackageType.DEVEL: None}
+        p = self.store.get_package('%s-pkg' % recipe.name)
+        tar = DistTarball(self.config, p, self.store)
+        # use the package (not the packager) to avoid the warnings
+        # TODO: separate runtime and devel packages after setting
+        # all files needed by each recipe properly
+        #ret[PackageType.RUNTIME] = tar.get_name(PackageType.RUNTIME)
+        ret[PackageType.DEVEL] = tar.get_name(PackageType.DEVEL)
+        return ret
+
+    def _apply_steps(self, recipe, steps, count, total):
+        self._ensure_ready(recipe)
+        for desc, step in steps:
+            m.build_step(count, total, recipe.name, step)
+            # check if the current step needs to be done
+            if self.cookbook.step_done(recipe.name, step) and not self.force:
+                m.action(_("Step done"))
+                continue
+
+            # check if the recipe was installed from a frozen one,
+            # in which case, we won't generate it again unless force is used
+            if step in [self.GEN_BINARY[1], self.UPLOAD_BINARY[1]]:
+                if self.cookbook.step_done(recipe.name, self.EXTRACT_BINARY[1]) and not self.force:
+                    m.action(_('No need since a package exists in remote for this recipe'))
+                    self.cookbook.update_step_status(recipe.name, step)
+                    continue
+
+            # call step function
+            stepfunc = getattr(self, step)
+            if not stepfunc:
```

Comment:
If it is possible that self has not attribute for step, you should use hasattr() before getattr() because if there is no self[step] getattr() will throw an AttributeError().

In this code your are checking that self[step] can be None or a function but you implicitly consider that step is always defined as an attribute of self.

Anyway, in all cases you will end up with a raised error, so result is the same.